src/puppetlabs/puppetdb/mq_listener.clj

```diff
@@ -330,64 +278,76 @@
    handler-fn :- message-fn-schema]
   (swap! listener-atom conj [pred handler-fn]))
 
-(defn start-receiver
-  [connection endpoint discard-dir process-msg]
-  (let [sess (.createSession connection true 0)
-        q (.createQueue sess endpoint)
-        consumer (.createConsumer sess q)
-        producer (.createProducer sess q)
-        send #(mq/commit-or-rollback sess
-                (.send producer (mq/to-jms-message sess %1 %2)))
-        handle (create-message-handler send discard-dir process-msg)]
-    (.setMessageListener
-     consumer
-     (reify MessageListener
-       (onMessage [this msg]
-         (try
-           (mq/commit-or-rollback sess (handle (mq/convert-jms-message msg)))
-           (catch Throwable ex
-             (log/error ex "message receive failed")
-             (throw ex))))))
-    {:session sess :consumer consumer :producer producer}))
+(def one-hour (* 1000 60 60))
+
+(defn send-delayed-message [command-chan delay-pool]
+  (fn [cmd]
+    (let [narrowed-entry (dissoc cmd :payload)]
+      (after one-hour #(async/>!! command-chan narrowed-entry) delay-pool))))
+
+(defn create-command-consumer
+  "Create and return a command handler. This function does the work of
+  consuming/storing a command. Handled commands are acknowledged here"
+  [q command-chan dlo delay-pool command-handler]
+  (let [handle-message (message-handler q
+                                        dlo
+                                        (send-delayed-message command-chan
+                                                              delay-pool)
+                                        command-handler)]
+    (fn [message]
+      ;; When the queue is shutting down, it sends nil message
+      (when message
+        (try
+          (handle-message message)
+          (catch Exception ex
+            (log/error ex "Unable to process message. Message not acknowledged and will be retried")))))))
+
+(defn create-command-handler-threadpool
+  "Creates an unbounded threadpool with the intent that access to the
+  threadpool is bounded by the semaphore. Implicitly the threadpool is
+  bounded by `size`, but since the semaphore is handling that aspect,
+  it's more efficient to use an unbounded pool and not duplicate the
+  constraint in both the semaphore and the threadpool"
+  [size]
+  (gtp/create-threadpool size "cmd-proc-thread-%d" 10000))
```

Comment:
This `10000` is an arbitrary number that could probably use some more thought. This is the number of milliseconds we will wait after telling the threadpool to shutdown before we start forcibly shutting down things.